### PR TITLE
Feature 150 highlight commits

### PR DESCRIPTION
**Pull request by aeogunmuyiwa**
Issue #150 adds a feature to the graph which allows users to distinguish between pushed commits (nodes) and un-pushed commits.

- Using simple-git, we first gather the commit IDs that are not on remote (using a command similar to git log), and then hash them. They are placed into an array which is then read in graphing.ts and graphSetup.ts. This allows us to "flag" nodes that are un-pushed, and then add a highlight ring around them in the graph space to alert the user.

- The graph should refresh automatically upon opening and after a push. Depending on the level of the graph you are viewing, you may need to zoom in to see the individual commit(s) that is/are highlighted.

Affected files:
graphing.ts
graphSetup.ts

Most affected functions:
populateCommits()

Note : local repo i.e. repo without upstream configured will show all node as normal (blue with highlight red).

### CODE REVIEW

**Comment by sterlinglaird** [comment]
When I tried this with branches that had no upstream (had never been pushed) I got the following error and the graph never loaded.

Error: fatal: no upstream configured for branch 'as'
    at C:\Users\Sterling\Desktop\project-virtual-team-1\node_modules\simple-git\promise.js:62
    at Git.<anonymous> (C:\Users\Sterling\Desktop\project-virtual-team-1\node_modules\simple-git\src\git.js:1556)
    at Function.Git.fail (C:\Users\Sterling\Desktop\project-virtual-team-1\node_modules\simple-git\src\git.js:1481)
    at fail (C:\Users\Sterling\Desktop\project-virtual-team-1\node_modules\simple-git\src\git.js:1439)
    at C:\Users\Sterling\Desktop\project-virtual-team-1\node_modules\simple-git\src\git.js:1448
    at ZoneDelegate.invoke (zone.js:323)
    at Zone.run (zone.js:216)
    at zone.js:571
    at ZoneDelegate.invokeTask (zone.js:356)
    at Zone.runTask (zone.js:256)
    at drainMicroTaskQueue (zone.js:474)
    at processTicksAndRejections (internal/process/task_queues.js:86)

**Comment by aeogunmuyiwa** [comment]
@sterlinglaird  Thanks for the review, I was wondering if you could run npm install and retry. That error should not persist .

**Comment by trstndvll** [comment]
> When I tried this with branches that had no upstream (had never been pushed) I got the following error and the graph never loaded.

Thanks for testing @sterlinglaird! Just to confirm, are you saying if you create a local repo and try to open it, the graph won't load? I myself have experienced that since the beginning of the course, so I just want to see if our changes caused it. Are you able to open local repos okay on other branches of the project?

**Comment by sterlinglaird** [comment]
@aeogunmuyiwa I git `npm install` but still get the issue.
@trstndvll I had an existing repo with remote branches, and then I made a branch that didn't have a remote (I dont know if you can do that with visual git, but it was an existing repo I made previously). If I have that local branch checked out, I get the error I mentioned, otherwise it seems to work fine.

**Comment by Unimax** [comment]
@trstndvll it looks like the error you mentioned also happen on master is it not? i just opened a new local repo on master and graph never loaded.

issues exits on the master are not part of this feature. let me know if this issue has caused a new issue or the feature itself not working.

**Comment by sterlinglaird** [comment]
The error I see doesn't happen in master and is introduced by these changes. As I said the repository is not new and has commits in it.  To reproduce I simply checkout a branch that has never been pushed to master. The branch that is checked out (has a head but no remote). Once I do that the graph doesn't load.

**Comment by Unimax** [comment]
> The error I see doesn't happen in master and is introduced by these changes. As I said the repository is not new and has commits in it. To reproduce I simply checkout a branch that has never been pushed to master. The branch that is checked out (has a head but no remote). Once I do that the graph doesn't load.
you are right. so, what is expected behavior all colored or none at all?

**Comment by Unimax** [comment]
@sterlinglaird fixed. can you test it again? the graph should load and the node should look normal.

**Comment by trstndvll** [comment]
We updated the PR description as well just to explain the fix: 

> Note : local repo i.e. repo without upstream configured will show all node as normal (blue with highlight red).

**Comment by sterlinglaird** [comment]
@Unimax The graph now loads, but now none of the highlights are shown for un-pushed commits when I checkout the local-only branch.

**Comment by Unimax** [comment]
> @Unimax The graph now loads, but now none of the highlights are shown for un-pushed commits when I checkout the local-only branch.

Right. added that as a note on the PR. that is expected.

**Comment by trstndvll** [comment]
> Right. added that as a note on the PR. that is expected.

@sterlinglaird We thought it would make the most sense that it only highlights un-pushed commits if there's a remote that actually exists. Otherwise there's going to be a lot of red on the screen (especially if you're in a situation where you delete your remote)!

**Comment by sterlinglaird** [comment]
With the new changes this is what it looks like if i checkout the local-only branch "as":
checked

vs checking out the branch "master" that exists on remote.
not-checked

In the second image the correct nodes are highlighted since they are not pushed. In the first image no highlights are shown, even though "master" merges the changes.

Shouldn't the same nodes be highlighted in both images?

**Comment by Athenasko** [comment]
While testing along the lines of @sterlinglaird, I tried to create a new local branch; however trying to zoom in causes the entire graph to disappear. Refreshing it or checking out a different branch before returning to it fixes it, however below is the error message that is displayed:

zone.js:461 Unhandled Promise rejection: Cannot add item: item with id 1 already exists ; Zone: <root> ; Task: Promise.then ; Value: Error: Cannot add item: item with id 1 already exists
    at DataSet._addItem (C:\Users\ancie\Docum…is\dist\vis.js:8842)
    at DataSet.add (C:\Users\ancie\Docum…is\dist\vis.js:8216)
    at makeNode (graphing.js:642)
    at graphing.js:177
    at new ZoneAwarePromise (zone.js:584)
    at populateCommits (graphing.js:107)
    at ZoneDelegate.invoke (zone.js:323)
    at Zone.run (zone.js:216)
    at zone.js:571
    at ZoneDelegate.invokeTask (zone.js:356)
    at Zone.runTask (zone.js:256)
    at drainMicroTaskQueue (zone.js:474)
    at ZoneTask.invoke (zone.js:426)
zone.js:463 Error: Uncaught (in promise): Error: Cannot add item: item with id 1 already exists
    at resolvePromise (zone.js:538)
    at resolvePromise (zone.js:523)
    at zone.js:571
    at ZoneDelegate.invokeTask (zone.js:356)
    at Zone.runTask (zone.js:256)
    at drainMicroTaskQueue (zone.js:474)
    at ZoneTask.invoke (zone.js:426)

**Comment by Unimax** [comment]
> With the new changes this is what it looks like if i checkout the local-only branch "as":
> checked
> 
> vs checking out the branch "master" that exists on remote.
> not-checked
> 
> In the second image the correct nodes are highlighted since they are not pushed. In the first image no highlights are shown, even though "master" merges the changes.
> 
> Shouldn't the same nodes be highlighted in both images?




> While testing along the lines of @sterlinglaird, I tried to create a new local branch; however trying to zoom in causes the entire graph to disappear. Refreshing it or checking out a different branch before returning to it fixes it, however below is the error message that is displayed:
> 
> zone.js:461 Unhandled Promise rejection: Cannot add item: item with id 1 already exists ; Zone: ; Task: Promise.then ; Value: Error: Cannot add item: item with id 1 already exists
> at DataSet._addItem (C:\Users\ancie\Docum…is\dist\vis.js:8842)
> at DataSet.add (C:\Users\ancie\Docum…is\dist\vis.js:8216)
> at makeNode (graphing.js:642)
> at graphing.js:177
> at new ZoneAwarePromise (zone.js:584)
> at populateCommits (graphing.js:107)
> at ZoneDelegate.invoke (zone.js:323)
> at Zone.run (zone.js:216)
> at zone.js:571
> at ZoneDelegate.invokeTask (zone.js:356)
> at Zone.runTask (zone.js:256)
> at drainMicroTaskQueue (zone.js:474)
> at ZoneTask.invoke (zone.js:426)
> zone.js:463 Error: Uncaught (in promise): Error: Cannot add item: item with id 1 already exists
> at resolvePromise (zone.js:538)
> at resolvePromise (zone.js:523)
> at zone.js:571
> at ZoneDelegate.invokeTask (zone.js:356)
> at Zone.runTask (zone.js:256)
> at drainMicroTaskQueue (zone.js:474)
> at ZoneTask.invoke (zone.js:426)

can you try to recreate the same on the master. let us know if it's only happening on our branch. because I don't remember making any changes which can give this error.

**Comment by Athenasko** [comment]
> > With the new changes this is what it looks like if i checkout the local-only branch "as":
> > checked
> > vs checking out the branch "master" that exists on remote.
> > not-checked
> > In the second image the correct nodes are highlighted since they are not pushed. In the first image no highlights are shown, even though "master" merges the changes.
> > Shouldn't the same nodes be highlighted in both images?
> 
> > While testing along the lines of @sterlinglaird, I tried to create a new local branch; however trying to zoom in causes the entire graph to disappear. Refreshing it or checking out a different branch before returning to it fixes it, however below is the error message that is displayed:
> > zone.js:461 Unhandled Promise rejection: Cannot add item: item with id 1 already exists ; Zone: ; Task: Promise.then ; Value: Error: Cannot add item: item with id 1 already exists
> > at DataSet._addItem (C:\Users\ancie\Docum…is\dist\vis.js:8842)
> > at DataSet.add (C:\Users\ancie\Docum…is\dist\vis.js:8216)
> > at makeNode (graphing.js:642)
> > at graphing.js:177
> > at new ZoneAwarePromise (zone.js:584)
> > at populateCommits (graphing.js:107)
> > at ZoneDelegate.invoke (zone.js:323)
> > at Zone.run (zone.js:216)
> > at zone.js:571
> > at ZoneDelegate.invokeTask (zone.js:356)
> > at Zone.runTask (zone.js:256)
> > at drainMicroTaskQueue (zone.js:474)
> > at ZoneTask.invoke (zone.js:426)
> > zone.js:463 Error: Uncaught (in promise): Error: Cannot add item: item with id 1 already exists
> > at resolvePromise (zone.js:538)
> > at resolvePromise (zone.js:523)
> > at zone.js:571
> > at ZoneDelegate.invokeTask (zone.js:356)
> > at Zone.runTask (zone.js:256)
> > at drainMicroTaskQueue (zone.js:474)
> > at ZoneTask.invoke (zone.js:426)
> 
> can you try to recreate the same on the master. let us know if it's only happening on our branch. because I don't remember making any changes which can give this error.

My bad. You're right @Unimax It is already a problem that can be experienced in master.

**Comment by Unimax** [comment]
@sterlinglaird nice catch

fixed will now show mark all unPushedCommit from all branches

**Comment by Unimax** [comment]
@sterlinglaird now it should be working. even when u are on a branch with no remote it will show all unpushed commits.

File: app/misc/graphing.ts [hunk 1]
```diff
@@ -36,6 +36,17 @@ let branchIds = [];
 let tagIds = [];
 let unumberPrev = 0;
 let selectedCommit: string;
+var aheadCommitList: any[] =[];
+
+
+/*
+Types of nodes in the network.
```
Comment (Unimax) [inline app/misc/graphing.ts]: idk why these comments are showing as part of our PR

File: app/misc/graphing.ts [hunk 2]
```diff
@@ -401,28 +426,33 @@ function makeBasicNode(c, column: number) {
...
-                fixed: false,
-                x: (column - 1) * spacingX,
-                y: (id - 1) * spacingY,
-                author: c.author(),
-                nodeType: NodeType.Basic
-            });
-
-        });        
 
+        console.log( title );
```
Comment (Unimax) [inline app/misc/graphing.ts]: Note some changes is due to formatting.

**Pull request merged**